### DISCUSSION TITLE
Bug about content length -1

### DISCUSSION THREAD

**Discussion by Nihilistttt** [corazawaf/coraza#875] [Q&A]
When response content length -1 that means header will send immediately. What if this request blocked that means response code will change 4xx or 5xx. But header already sent. It will not change antmore.

**Comment by jcchavezs** [comment]
Thanks @Nihilistttt in what connector are you experiencing this?

**Reply by Nihilistttt** [reply]
Connector?

**Reply by Nihilistttt** [reply]
func (i *rwInterceptor) WriteHeader(statusCode int) {
	if i.wroteHeader {
		log.Println("http: superfluous response.WriteHeader call")
		return
	}

	for k, vv := range i.w.Header() {
		for _, v := range vv {
			i.tx.AddResponseHeader(k, v)
		}
	}

	i.statusCode = statusCode
	if it := i.tx.ProcessResponseHeaders(statusCode, i.proto); it != nil {
		i.Header().Set("Content-Length", "0")
		i.statusCode = obtainStatusCodeFromInterruptionOrDefault(it, i.statusCode)
		i.flushWriteHeader()
		return
	}

	i.wroteHeader = true
}

In caddy repository interceptor.go file same with others. What will happen when reponse content-length is not exists or set to -1. After processing if content-length is -1. Server.go try to send header immediately. And if the response needs to be blocked, cant be done any more because response header already sent to client. Also in rwinterceptor element names w not holding right status code. I wrote example application for it. I didnt know the problem at that time. You can try it 
https://github.com/corazawaf/coraza/discussions/874

**Comment by jptosso** [comment]
Coraza doesn't perform any validation regarding this behavior, it's up to the ruleset to do so. It will also depend on the behavior of the connector, for example, if you use caddy, caddy will receive a chunked body and coraza will receive the real content length and the full body buffered.

**Reply by Nihilistttt** [reply]
I am using just regulat coraza package in reverse proxy what I explained before this.

**Reply by jptosso** [reply]
Can you share more details?

**Reply by Nihilistttt** [reply]
func (i *rwInterceptor) WriteHeader(statusCode int) {
if i.wroteHeader {
log.Println("http: superfluous response.WriteHeader call")
return
}

for k, vv := range i.w.Header() {
	for _, v := range vv {
		i.tx.AddResponseHeader(k, v)
	}
}

i.statusCode = statusCode
if it := i.tx.ProcessResponseHeaders(statusCode, i.proto); it != nil {
	i.Header().Set("Content-Length", "0")
	i.statusCode = obtainStatusCodeFromInterruptionOrDefault(it, i.statusCode)
	i.flushWriteHeader()
	return
}

i.wroteHeader = true
}

In caddy repository interceptor.go file same with others. What will happen when reponse content-length is not exists or set to -1. After processing if content-length is -1. Server.go try to send header immediately. And if the response needs to be blocked, cant be done any more because response header already sent to client. Also in rwinterceptor element names w not holding right status code. I wrote example application for it. I didnt know the problem at that time. You can try it
https://github.com/corazawaf/coraza/discussions/874
